### DISCUSSION TITLE
Resolving Cross-references works in Test but not in running editor

### DISCUSSION THREAD

**Discussion by tbuchmann** [eclipse-langium/langium#1781] [Q&A] [answered]
Hello everybody,

I'm switching from Xtext to Langium and so far everything works fine (despite I'm still learning TypeScript). However, I ran into a strange problem where I need some help.

I want to write a small language for modeling UML class diagrams with a textual syntax.
Types (Class | Interface | Enum | DataType | PrimitiveType | Association) are contained in packages. Classes may extend other Classes (and implement Interfaces), whereas Interfaces only may extend other Interfaces. A cutout of the grammar looks like this:

```
type Type =
    Class | Interface | DataType | PrimitiveType | Enumeration | Association;

entry Model:
    packages+=Package*;

Package:
    'package' name=ID '{'
        (packages+=Package | types+=(Class|Interface|DataType|PrimitiveType|Enumeration|Association) )*
    '}';

Interface returns Type:
    {infer Interface} (vis=VisibilityKind)? 'interface' name=ID 
    ('extends' superInterfaces+=[Interface:ID] (',' superInterfaces+=[Interface:ID])*)? '{'
        (properties+=Property | operations+=Operation)*
    '}';
```

I'm facing the following problem: The following test works as expected

```
test('parse interface with multiple inheritance', async () => {
        document = await parse(`
            package de {
                interface A {}
                interface B {}
                interface C extends A, B {}
            }
        `);        

        expect(document.parseResult.parserErrors).toHaveLength(0);

        expect(
            checkDocumentValid(document) || s`
            Interfaces:
              ${(document.parseResult.value?.packages?.[0].types as Interface[]).map(c => c.name)?.join('\n')}
            Super-Interfaces:
              ${(document.parseResult.value?.packages?.[0].types?.[2] as Interface).superInterfaces?.map(i => i.ref?.name)?.join('\n')}
            `
        ).toBe(s`
            Interfaces:
              A
              B
              C
            Super-Interfaces:
              A
              B
        `);
    });
```

But when I start the extension in debugging mode and enter the same text into a new document the superInterfaces reference of interface "C" is empty.

**Comment by cdietrich** [comment]
How do you evaluate this in your extension? Refs are getters so you need to call them

**Reply by tbuchmann** [reply]
Ah sorry, my post was cut short:

I noticed the problem when debugging the code generator. I observed that only code for the declaration of the interfaces was generated in my target language. After setting a breakpoint in the generator.ts file I was able to see the content of the model. Navigating to the package and the interfaces showed that the superTypes reference for Interface C was empty.

**Reply by cdietrich** [reply]
How do you implement/ call the generator

**Reply by tbuchmann** [reply]
I added a listener (vscode.workspace.onDidSaveTextDoument) to extension/main.ts

In the listener, I basically do the same steps as in cli/main.ts.
Other cross references (e.g. Property Types) are resolved fine btw...

**Reply by cdietrich** [reply]
Hmm. Then I guess you need to debug the linker

**Reply by cdietrich** [reply]
Maybe you can also provide a reproducer

**Reply by tbuchmann** [reply]
I could give you access to the repository, where the code is hosted if that helps. 
Steps to reproduce: Run parsing.test.ts (works)
Set breakpoint in line 146 of generator.ts
Press F5 to start extension in Debug mode, create a new file with ending .cdiag
Copy the following text to the new file:
```
package de {
    interface A {}
    interface B {}
    interface C extends A, B {}
}
```
Save the file (triggers the generator).

**Reply by cdietrich** [reply]
Yes I can have a look later today

**Reply by cdietrich** [reply, marked as answer]
the problem is your
checkNoCycleInInterfaceInheritance
the pops might also pop them them their original place...

**Reply by tbuchmann** [reply]
Oh, you're right. Thanks a lot for your time and help!
Guess that calls for some integration tests ;)
